components/brave_perf_predictor/browser/third_party_extractor.cc

```diff
@@ -0,0 +1,148 @@
+
+ThirdPartyExtractor::ThirdPartyExtractor() = default;
+
+bool ThirdPartyExtractor::InitializeFromResource() {
+  const auto resource_id = IDR_THIRD_PARTY_ENTITIES;
+
+  auto& resource_bundle = ui::ResourceBundle::GetSharedInstance();
+  std::string data_resource;
+  if (resource_bundle.IsGzipped(resource_id)) {
+    data_resource = resource_bundle.DecompressDataResource(resource_id);
+  } else {
+    data_resource = resource_bundle.GetRawDataResource(resource_id).as_string();
+  }
+
+  return LoadEntities(data_resource);
+}
+
+bool ThirdPartyExtractor::LoadEntities(const std::string& entities) {
+  // Reset previous mappings
+  entity_by_domain_.clear();
+  entity_by_root_domain_.clear();
+
+  // Parse the JSON
+  base::Optional<base::Value> document = base::JSONReader::Read(entities);
+  if (!document || !document->is_list()) {
+    LOG(ERROR) << "Bad Document";
+    return false;
+  }
+  base::ListValue* entities_parsed = nullptr;
+  if (!document->GetAsList(&entities_parsed))
+    return false;
+
+  // Collect the mappings
+  for (auto& entity : *entities_parsed) {
+    base::DictionaryValue* entity_dict = nullptr;
+    if (!entity.GetAsDictionary(&entity_dict))
+      continue;
+    auto* entity_name = entity_dict->FindKey("name");
+    auto* entity_domains = entity_dict->FindKey("domains");
+    if (!entity_name || !entity_domains)
+      continue;
+    std::string entity_name_string;
+    if (!entity_name->GetAsString(&entity_name_string))
+      continue;
+    base::ListValue* entity_domains_list = nullptr;
+    if (!entity_domains->GetAsList(&entity_domains_list))
+      continue;
+
+    for (auto& entity_domain : *entity_domains_list) {
+      std::string entity_domain_string;
+      if (!entity_domain.GetAsString(&entity_domain_string))
+        continue;
+
+      auto entity_entry = entity_by_domain_.find(entity_domain_string);
+      if (entity_entry != entity_by_domain_.end()) {
+        LOG(ERROR) << "Duplicate domain " << entity_domain_string;
+      } else {
+        entity_by_domain_.emplace(entity_domain_string, entity_name_string);
+        auto root_domain = GetRootDomain(entity_domain_string);
+
+        auto root_entity_entry = entity_by_root_domain_.find(root_domain);
+        if (root_entity_entry != entity_by_root_domain_.end() &&
+            root_entity_entry->second != entity_name_string) {
+          // If there is a clash at root domain level, neither is correct
+          entity_by_root_domain_.erase(root_entity_entry);
+        } else {
+          entity_by_root_domain_.emplace(root_domain, entity_name_string);
+        }
+      }
+    }
+  }
+
+  initialized_ = true;
+
+  return true;
+}
+
+ThirdPartyExtractor::~ThirdPartyExtractor() = default;
+
+// static
+ThirdPartyExtractor* ThirdPartyExtractor::GetInstance() {
+  auto* extractor = base::Singleton<ThirdPartyExtractor>::get();
+  // By default initialize from packaged resources
+  if (!extractor->IsInitialized())
+    extractor->InitializeFromResource();
```

Comment:
I'm not happy with doing this initialization on the main thread during the URL load - parsing of 65kb json is absolutely not an invisible operation (literally) - we definitely will loose a bunch of frames on UI, especially on mobiles. Since this is not something mission critical, I'd suggest to offload this init to something background.

To do this, and for the sake of sanity I'd also suggest to get rid of singleton by handing off this instance to, for example, browser process and doing a delayed initialization there